Blog: python copy-paste shutil methods (#120)
JankariTech/blog: src/assets/PythonCopyFileWithShutil/python-copy-paste.md

---

> title: Ways to copy file with shutil module in Python

Reviewer: ```suggestion
title: Ways to copy a file with the shutil module in Python
```

authorName: Salipa Gurung
authorAvatar: https://avatars.githubusercontent.com/u/53458341?v=4
authorLink: https://github.com/Salipa-Gurung
createdAt: March 11, 2024
tags: python, copy, shutil
banner: https://blog.jankaritech.com/src/assets/PythonCopyFileWithShutil/images/python_copy_file.png
---

> Python provides many built-in modules to copy a file. We'll look into ways to copy file using `shutil` module in this blog post.

Reviewer: ```suggestion
Python provides many built-in modules to copy a file. We'll look into ways to copy a file using the `shutil` module in this blog post.
```

> The standard utility module known as shutil is an abbreviation for shell utilities. Numerous functions in the Python shutil module helps with high-level file and directory operations. Using the shutil module, there are four common methods that provide an easy and efficient way to copy a file:

Reviewer: ```suggestion
The standard utility module known as shutil is an abbreviation for shell utilities. Numerous functions in the Python shutil module help with high-level file and directory operations. Using the shutil module, there are four common methods that provide an easy and efficient way to copy a file:
```

### 1. shutil.copy() method

```py
shutil.copy(source, destination, *, follow_symlinks=True)
```

> This method copies the file of source to the directory or file specified in the destination and returns file's destination.

Reviewer: ```suggestion
This method copies the source file to the directory or file specified in the destination and returns the file's destination.
```
Maybe "destination path" - I suppose it returns some sort of path with directory(ies) and file name.

- shutil.copy() method copies file's data and file's permission mode.
- Metadata(file creation and modification time) of the file copied is not preserved.
- Source and destination should be path-like objects or strings.

> - Destination can specify both file or directory. If the destination specify path to a directory, the file will be copied to destination directory with the basename of the source path.

Reviewer: ```suggestion
- Destination can specify both file or directory. If the destination specifies a path to a directory, the file will be copied to the destination directory with the basename of the source path.
```

Is "basename" OK here, or too technical a term. Could just say "filename of the source path"?

- Copying symbolic link:

>   - If `follow_symlinks=True`, destination will be a copy of the file that the symbolic link in source points to. Default value of parameter `follow_symlinks` is `True`.

Reviewer: ```suggestion
  - If `follow_symlinks=True`, the destination will be a copy of the file that the symbolic link in source points to. The default value of parameter `follow_symlinks` is `True`.
```

>   - If `follow_symlinks=False`, destination will be created as symbolic link.

Reviewer: ```suggestion
  - If `follow_symlinks=False`, destination will be created as a symbolic link.
```

  ```py
  # Import shell utility module
  import shutil

  #  Copy the content of source_file.txt to destination_file.txt
  shutil.copy('source_file.txt', 'destination_file.txt')
  ```

### 2. shutil.copy2() method

```py
shutil.copy2(source, destination, *, follow_symlinks=True)
```

> The shutil.copy2() method provides additional functionality of preserving all the metadata of the file it can. Other characteristics are same as shutil.copy() method.

Reviewer 2: ```suggestion
This method provides additional functionality of preserving all the metadata of the file it can. Other characteristics are same as shutil.copy() method.
```

  - shutil.copy2() method copies file's data and file's permission mode.
  - Source and destination should be path-like objects or strings.

>   - Destination can specify both file or directory. If the destination is directory, the file will be copied with the same name as of the source file name.

Reviewer: ```suggestion
  - Destination can specify both file or directory. If the destination is a directory, the file will be copied with the same name as the source file name.
```

    ```py
    # Import shell utility module
    import shutil

    #  Copy the content of source_file.txt to destination_file.txt
    shutil.copy2('source_file.txt', 'destination_file.txt')
    ```

### 3. shutil.copyfile() method

```py
shutil.copyfile(source, destination, *, follow_symlinks=True)
```
Using this method, a file's content is copied from its source to its destination without any metadata.

  - Source and destination should be path-like objects or strings.

>   - Destination path can only specify path to a file but cannot specify path to a directory.

Reviewer: ```suggestion
  - Destination path can only specify a path to a file but cannot specify a path to a directory.
```

    ```py
    # Import shell utility module
    import shutil

    #  Copy the content of source_file.txt to destination_file.txt
    shutil.copyfile('source_file.txt', 'destination_file.txt')
    ```

### 4. shutil.copyfileobj() method

```py
shutil.copyfileobj(fsrc, fdst, length)
```

- Source and destination must be file-like objects.

> - This method copies contents of file object to the destination file object.

Reviewer: ```suggestion
- This method copies contents of the source file object to the destination file object.
```

> - Source and destination must be file-like objects.
> - This method copies contents of file object to the destination file object.

Reviewer 2: ```suggestion
- This method copies contents of file object to the destination file object.
- Source and destination must be file-like objects.
```

> - Length parameter which is optional, specifies an integer value for buffer size.

Reviewer: ```suggestion
- Length parameter, which is optional, specifies an integer value for buffer size.
```

This seems strange, why is the parameter called "length" but controls "buffer size"? Is that just "the way it is" in python?

- It does not preserve metadata
- It does not return any value.

  ```py
  # Import shell utility module
  import shutil

  # Create file objects
  source_file_object = open("../hello.txt", "r")
  destination_file_object = open("CopyFile/copyFalseSrcPath.txt", "w")

  # Copy content of file object to another file object
  shutil.copyfileobj(source_file_object, destination_file_object)
  ```

### Points to prevent common mistakes:

>  - The source and destination cannot specify the same path in shutil.copy(), shutil.copy2() and shutil.copyfile(). If source and destination specify same file path then exception `SameFileError` will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file file if it already exists.

Reviewer: ```suggestion
 - The source and destination cannot specify the same path in shutil.copy(), shutil.copy2() and shutil.copyfile(). If the source and destination specify the same file path then exception `SameFileError` will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file if it already exists.
```

>  - The source and destination cannot specify the same path in shutil.copy(), shutil.copy2() and shutil.copyfile(). If source and destination specify same file path then exception `SameFileError` will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file file if it already exists.

Reviewer 2: ```suggestion
 - The source and destination cannot specify the same path in `shutil.copy()`, `shutil.copy2()` and `shutil.copyfile()`. If source and destination specify same file path then exception "SameFileError" will be raised. This error is not raised in `shutil.copyfileobj()` as it overwrites the file file if it already exists.
```

I think this will look better

Reviewer 3: good point, @nabim777 , it's easier for the reader to understand the names of the actual commands this way!

>  - The source and destination cannot specify the same path in `shutil.copy()`, `shutil.copy2()` and `shutil.copyfile()`. If the source and destination specify the same file path then exception "SameFileError" will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file file if it already exists.

Reviewer: ```suggestion
 - The source and destination cannot specify the same path in `shutil.copy()`, `shutil.copy2()` and `shutil.copyfile()`. If the source and destination specify the same file path then exception "SameFileError" will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the destination file if it already exists.
```

>  - An error `FileNotFound` will be raised if the source specifies a path to a file that is not present.

Reviewer 2: ```suggestion
 - An error "FileNotFound" will be raised if the source specifies a path to a file that is not present.
```

---
title: Ways to copy file with shutil module in Python
authorName: Salipa Gurung
authorAvatar: https://avatars.githubusercontent.com/u/53458341?v=4
authorLink: https://github.com/Salipa-Gurung

> createdAt: Feb 27, 2024

Reviewer 3: Set date to the date when it will be published

tags: python, copy-paste, shutil
banner: https://blog.jankaritech.com/src/assets/PythonCopyFileWithShutil/images/python_copy_file.png
---

> Python provides many built-in modules to copy and paste a file. We'll look into ways to copy file using `shutil` module in this blog post.

Reviewer 4: Is it called `copy and paste` when done programmatically, or is it just `copy` a file

> Python provides many built-in modules to copy and paste a file. We'll look into ways to copy file using `shutil` module in this blog post.

Reviewer 3: As a reader I would be interested why shutil is recommended by you since you mention that there are many modules available for copy and pasting of files. Maybe mention one or two things that are particularly nice about shutil

> The standard utility module known as shutil is an abbreviation for shell utilities. Numerous functions in the Python shutil module helps with high-level file and directory operations. Using the shutil module, there are four common methods that provide an easy and efficient way to copy and paste a file:

Reviewer 3: ```suggestion
The standard utility module known as shutil is an abbreviation for shell utilities. Numerous functions in the Python shutil module help with high-level file and directory operations. Using the shutil module, there are four common methods that provide an easy and efficient way to copy and paste a file:
```

because the subject (numerous functions) is in plural

> The standard utility module known as shutil is an abbreviation for shell utilities. Numerous functions in the Python shutil module helps with high-level file and directory operations. Using the shutil module, there are four common methods that provide an easy and efficient way to copy and paste a file:

Reviewer 5: What does `high-level` mean in this context?

### 1. shutil.copy() method

```py
shutil.copy(source, destination, *, follow_symlinks=True)
```

This method copies the file of source to the directory or file specified in the destination and returns file's destination.

> 

Reviewer 3: If you write the following section in the bullet point style, it would be nice to have an introductory sentence, something like "This method is characterized by the following points" or "Let's have a look at the parameters of the method"

- shutil.copy() method copies file's data and file's permission mode.

> - Metadata(file creation and modification time) of the file copied is not preserved.

Reviewer 3: ```suggestion
- Metadata (file creation and modification time) of the file copied is not preserved.
```

- Source and destination should be path-like objects or strings.

> - Destination can specify both file or directory. If the destination is directory, the file will be copied with the same name as of the source file name.

Reviewer 4: the name will be the same, but where will be the file? Inside of the destination directory?

Reviewer 3: resolving this since it has been addressed in the meanwhile

> - Copying symbolic link:

Reviewer 4: A list should have some kind of "heading" that indicates what the list is about.

Food I love:
- momo
- biryani
- ...

  - If `follow_symlinks=True`, destination will be a copy of the file that the symbolic link in source points to. Default value of parameter `follow_symlinks` is `True`.

  - If `follow_symlinks=False`, destination will be created as symbolic link.

  ```py
  # Import shell utility module
  import shutil

  #  Copy the content of source_file.txt to destination_file.txt
  shutil.copy('source_file.txt', 'destination_file.txt')
  ```

### 2. shutil.copy2() method

```py
shutil.copy2(source, destination, *, follow_symlinks=True)
```

> The shutil.copy2() method provides additional functionality of preserving all the metadata of the file it can. Other characteristics are same as shutil.copy() method.

Reviewer 3: ```suggestion
The shutil.copy2() method provides additional functionality of preserving all the metadata of the file it can. Other characteristics are same as for the shutil.copy() method.
```

I think you need a preposition here...

  - shutil.copy2() method copies file's data and file's permission mode.
  - Source and destination should be path-like objects or strings.

>   - Destination can specify both file or directory. If the destination is directory, the file will be copied with the same name as of the source file name.

Reviewer 4: same as above, what is the list about?

  - Copying symbolic link:

    - If `follow_symlinks=True`, destination will be a copy of the file that the symbolic link in source points to. Default value of parameter `follow_symlinks` is `True`.

    - If `follow_symlinks=False`, destination will be created as symbolic link.

    ```py
    # Import shell utility module
    import shutil

    #  Copy the content of source_file.txt to destination_file.txt
    shutil.copy2('source_file.txt', 'destination_file.txt')
    ```

### 3. shutil.copyfile() method

```py
shutil.copyfile(source, destination, *, follow_symlinks=True)
```
Using this method, a file's content is copied from its source to its destination without any metadata.

  - Source and destination should be path-like objects or strings.
  - Destination path can only specify path to a file but cannot specify path to a directory.
  - Copying symbolic link:

    - If `follow_symlinks=True`, destination will be a copy of the file that the symbolic link in source points to. Default value of parameter `follow_symlinks` is `True`.

>     - If `follow_symlinks=False`, destination will be created as symbolic link.

Reviewer 4: That seems to be the same for all cases, so maybe it would be enough to mention it once

Reviewer 3: Explain the things that are different between these methods. As a reader I would also be interested to learn in which situation which method is recommended.

    ```py
    # Import shell utility module
    import shutil

    #  Copy the content of source_file.txt to destination_file.txt
    shutil.copyfile('source_file.txt', 'destination_file.txt')
    ```

### 4. shutil.copyfileobj() method

```py
shutil.copyfileobj(fsrc, fdst, length)
```

- Source and destination must be file-like objects.
- This method copies contents of file object to the destination file object.
- Length parameter which is optional, specifies an integer value for buffer size.
- It does not preserve metadata
- It does not return any value.

  ```py
  # Import shell utility module
  import shutil

  # Create file objects
  source_file_object = open("../hello.txt", "r")
  destination_file_object = open("CopyFile/copyFalseSrcPath.txt", "w")

  # Copy content of file object to another file object
  shutil.copyfileobj(source_file_object, destination_file_object)
  ```

> Note:
> - The source and destination cannot specify the same path. If source and destination specify same file path then exception `SameFileError` will be raised.

> > - An error `FileNotFound` will be raised if the source specifies a path to a file that is not present.

Reviewer 4: Is that the case for all functions or only for the last? If its for all, it should be in a common section

Using this method, a file's content is copied from its source to its destination without any metadata.

- Source and destination should be path-like objects or strings.
- Destination path can only specify a path to a file but cannot specify a path to a directory.

  ```py
  # Import shell utility module
  import shutil

  #  Copy the content of source_file.txt to destination_file.txt
  shutil.copyfile('source_file.txt', 'destination_file.txt')
  ```

### 4. shutil.copyfileobj() method

```py
shutil.copyfileobj(fsrc, fdst, length)
```

- Source and destination must be file-like objects.
- This method copies contents of the source file object to the destination file object.
- Length parameter, which is optional, specifies an integer value for buffer size.
- It does not preserve metadata
- It does not return any value.

  ```py
  # Import shell utility module
  import shutil

  # Create file objects
  source_file_object = open("../hello.txt", "r")
  destination_file_object = open("CopyFile/copyFalseSrcPath.txt", "w")

  # Copy content of file object to another file object
  shutil.copyfileobj(source_file_object, destination_file_object)
  ```

### Points to prevent common mistakes:

> - The source and destination cannot specify the same path in `shutil.copy()`, `shutil.copy2()` and `shutil.copyfile()`. If the source and destination specify the same file path then exception "SameFileError" will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file file if it already exists.

Reviewer 2: ```suggestion
- The source and destination cannot specify the same path in `shutil.copy()`, `shutil.copy2()` and `shutil.copyfile()`. If the source and destination specify the same file path then exception "SameFileError" will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file if it already exists.
```

---
title: Python's `shutil` module for Automated Testing
authorName: Salipa Gurung
authorAvatar: https://avatars.githubusercontent.com/u/53458341?v=4
authorLink: https://github.com/Salipa-Gurung

> createdAt: March 11, 2024

Reviewer 3: update date before publishing

tags: python, copy, shutil
banner: https://blog.jankaritech.com/src/assets/PythonCopyFileWithShutil/images/python_copy_file.png
---

## Introduction

In the dynamic domain of software development, ensuring application quality through automated testing is vital. Automated testing involves running numerous tests that require a specific setup of files and directories. Setting up and tearing down test environments efficiently is a significant part of this process which can often be a tedious and error-prone task without the right tools.

> Finding the right tools to simplify tasks can make a lot of difference. The standard utility module known as shutil is an abbreviation for shell utilities. It provides high-level file operations for copying, moving, and removing files and directories that can simplify managing test environments. This blog explains how the shutil module can streamline file operations in automated software testing.

Reviewer 3: ```suggestion
Finding the right tools to simplify tasks can make a big difference. The standard utility module known as shutil is an abbreviation for shell utilities. It provides high-level file operations for copying, moving, and removing files and directories that can simplify managing test environments. This blog explains how the shutil module can streamline file operations in automated software testing.
```
I think it's more common to say "a big difference" than "a lot of difference". Maybe ask the English teacher what is the best term in this sentence?

To ensure that each test starts with the correct environment we need to automate the setup process creating a consistent and isolated test environment.

## Getting started with `shutil`
Shutil module is part of the Python Standard Library, so we don’t need to install anything extra. Simply import it at the beginning of your script.

```py
# Import shell utility module
import shutil
```

> ### Essential file copying methods

Reviewer 3: Do you think this point should really be nested under the previous point?
 
```suggestion
## Essential file copying methods
```

While testing an application we often require a specific set of data files. Instead of manually copying files, four common methods provide an easy and efficient way to copy a file using the shutil module.

1. [shutil.copy()](#1-shutilcopy-method)
2. [shutil.copy2()](#2-shutilcopy2-method)
3. [shutil.copyfile()](#3-shutilcopyfile-method)
4. [shutil.copyfileobj()](#4-shutilcopyfileobj-method)

> #### 1. shutil.copy() method

Reviewer 3: depending on what you decide about the title above, the following subtitles need to be adjusted
```suggestion
### 1. shutil.copy() method
```

```py
shutil.copyfile(source, destination, *, follow_symlinks=True)
```

Using this method, a file's content is copied from its source to its destination without metadata. The source and destination should be path-like objects or strings. Destination path can only specify a path to a file but cannot specify a path to a directory.

```py
# Import shell utility module
import shutil

#  Copy the content of source_file.txt to destination_file.txt
shutil.copyfile('source_file.txt', 'destination_file.txt')
```

#### 4. shutil.copyfileobj() method

```py
shutil.copyfileobj(fsrc, fdst, length)
```

This method copies the contents of the source file object to the destination file object. The source and destination must be file-like objects. It does not preserve metadata and does not return any value. The optional length parameter specifies an integer value for buffer size.

```py
# Import shell utility module
import shutil

# Create file objects
source_file_object = open("../hello.txt", "r")
destination_file_object = open("CopyFile/copyFalseSrcPath.txt", "w")

# Copy the content of the file object to another file object
shutil.copyfileobj(source_file_object, destination_file_object)
```

> ### Preventing common mistakes:

Reviewer 3: ```suggestion
## Preventing common mistakes:
```

- The source and destination cannot specify the same path in `shutil.copy()`, `shutil.copy2()` and `shutil.copyfile()`. If the source and destination specify the same file path then the exception "SameFileError" will be raised. This error is not raised in shutil.copyfileobj() as it overwrites the file if it already exists.
- An error "FileNotFound" will be raised if the source specifies a path to a file that is not present.

> ### Conclusion

Reviewer 3: ```suggestion
## Conclusion
```

Each method provides different aspects of file operations, providing flexibility and power to meet various testing needs. Python's `shutil` module offers essential tools for efficient file operations in software testing:

- `shutil.copy`: Great for basic file copying, including permissions.
- `shutil.copy2`: Ideal for copying files along with most of their metadata, including timestamps.
- `shutil.copyfile`: Best for fast, simple content-copying without metadata.
- `shutil.copyfileobj`: Enables precise copying between file-like objects.

Some other commonly used utility functions that can be helpful while automating our tests are:

1. `shutil.mkdirs()`: Creates directories recursively. Useful for setting up directory structures needed for testing or organizing files.
2. `shutil.copytree()`: Recursively copies an entire directory tree. Ideal for duplicating complex directory structures, including all files and sub-directories, to create or backup environments for testing or deployment.

> 3. `shutil.rmtree()`: Deletes a directory tree recursively.Essential for cleaning up after tests or removing temporary directories and their contents completely.

Reviewer 3: ```suggestion
3. `shutil.rmtree()`: Deletes a directory tree recursively. Essential for cleaning up after tests or removing temporary directories and their contents completely.
```

4. `shutil.move()`: Moves a file or directory to another location. Efficiently relocates files or directories within the filesystem, facilitating organization or deployment process.

> 5. `shutil.make_archive()`: Creates an archieve(such as AIP or TAR) of a directory and it's contents. Useful for packaging directories for distributions or backup purposes.

Reviewer 3: ```suggestion
5. `shutil.make_archive()`: Creates an archieve (such as AIP or TAR) of a directory and its contents. Useful for packaging directories for distributions or backup purposes.
```
I think there is a difference in English and Nepali about spaces with punctuation marks like `()`, `:` and similar because I have noticed that many don't add a space before the bracket. In English, there is always a space on the outside of the brackets, but not on the inside. You can also ask the English teacher about this if you want, I'm sure she can explain this better than I.

---
title: Python's `shutil` module for Automated Testing
authorName: Salipa Gurung
authorAvatar: https://avatars.githubusercontent.com/u/53458341?v=4
authorLink: https://github.com/Salipa-Gurung
createdAt: March 11, 2024
tags: python, copy, shutil
banner: https://blog.jankaritech.com/src/assets/PythonCopyFileWithShutil/images/python_copy_file.png
---

## Introduction

In the dynamic domain of software development, ensuring application quality through automated testing is vital. Automated testing involves running numerous tests that require a specific setup of files and directories. Setting up and tearing down test environments efficiently is a significant part of this process which can often be a tedious and error-prone task without the right tools.

Finding the right tools to simplify tasks can make a big difference. The standard utility module known as shutil is an abbreviation for shell utilities. It provides high-level file operations for copying, moving, and removing files and directories that can simplify managing test environments. This blog explains how the shutil module can streamline file operations in automated software testing.

> To ensure that each test starts with the correct environment we need to automate the setup process creating a consistent and isolated test environment.

Reviewer 6: ```suggestion
To ensure that each test starts with the correct environment, we need to automate the setup process, creating a consistent and isolated test environment.
```

## Getting started with `shutil`

Shutil module is part of the Python Standard Library, so we don’t need to install anything extra. Simply import it at the beginning of your script.

```py
# Import shell utility module
import shutil
```

## Essential file copying methods

> While testing an application we often require a specific set of data files. Instead of manually copying files, four common methods provide an easy and efficient way to copy a file using the shutil module.

Reviewer 6: ```suggestion
While testing an application, we often require a specific set of data files. Instead of manually copying files, four common methods provide an easy and efficient way to copy a file using the shutil module.
```

- `shutil.copy`: Great for basic file copying, including permissions.
- `shutil.copy2`: Ideal for copying files along with most of their metadata, including timestamps.
- `shutil.copyfile`: Best for fast, simple content-copying without metadata.
- `shutil.copyfileobj`: Enables precise copying between file-like objects.

Some other commonly used utility functions that can be helpful while automating our tests are:

1. `shutil.mkdirs()`: Creates directories recursively. Useful for setting up directory structures needed for testing or organizing files.
2. `shutil.copytree()`: Recursively copies an entire directory tree. Ideal for duplicating complex directory structures, including all files and sub-directories, to create or backup environments for testing or deployment.
3. `shutil.rmtree()`: Deletes a directory tree recursively. Essential for cleaning up after tests or removing temporary directories and their contents completely.
4. `shutil.move()`: Moves a file or directory to another location. Efficiently relocates files or directories within the filesystem, facilitating organization or deployment process.

> 5. `shutil.make_archive()`: Creates an archieve (such as AIP or TAR) of a directory and its contents. Useful for packaging directories for distributions or backup purposes.

Reviewer 6: I think you mean "archive"?

More utility functions and detailed descriptions can be found in the [official documentation](https://docs.python.org/3/library/shutil.html).

> The `shutil` module in Python module serves as a versatile toolset not only in software testing but also across diverse fields of software development such as version control and deployment, data science and machine learning, IT and system administration, research and academia, and many more.

Reviewer 6: ```suggestion
The `shutil` module in Python serves as a versatile toolset not only in software testing but also across diverse fields of software development such as version control and deployment, data science and machine learning, IT and system administration, research and academia, and many more.
```

> Integrating `shutil` into our testing processes we can perform complex file operations with minimal code, saving our time and effort and allowing us to focus on developing high-quality software.

Reviewer 6: Maybe:
```suggestion
Integrating `shutil` into our testing processes, we can perform complex file operations with minimal code, saving our time and effort and allowing us to focus on developing high-quality software.
```
?
